orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy

```diff
@@ -204,6 +205,52 @@ class OperationsController {
     }
   }
 
+  private void addStageAuthorizedRoles(def request, Map pipeline) {
```

Comment:
I don't understand what this is doing, or why it's doing it when we're starting a pipeline.

Is the idea not that a manual judgment stage can be configured with a set of roles needed to judge it?

Reply:
This method will check for manual judgment stages and will find out if the user is authorized to continue. Will set the isAuthorized flag to true , if authorized,

Reply:
It's still unclear why it needs to happen in the `OperationsController`. I'm confused as this code seems to be manipulating every stage in the pipeline?

Overall this implementation is hard to follow.

My expectation was that the manual judgment stage would have a set of required roles in its context, and would check against those roles when determining whether or not to continue. That shouldn't require changes anywhere other than the manual judgement stage/task.

As an aside, there is a small loophole here in that any stage context can be PATCH'd. See https://github.com/spinnaker/orca/blob/master/orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy#L475.

Reply:
It doesn't manipuate every stage. It only manipulates manual judgment stages.

Please refer the design document.
https://github.com/spinnaker/governance/pull/80

Reply:
```
def stageList = pipeline.stages
            def stageRoles = []
            stageList.each { item ->
              stageRoles = item.selectedStageRoles
              item.isAuthorized = ManualJudgmentAuthzGroupsUtil.checkAuthorizedGroups(userRoles, stageRoles, permissions)
              item.stageRoles = stageRoles
              item.permissions = permissions
            }
```

What am I missing here, that's iterating across `pipeline.stages`.

Reply:
item.selectedStageRoles returns null for all the stages except manual judgment(if configured).
This decision to apply for all stages was taken in our design meet, so that in future, if we want to extend it to all stages, this will work.

Reply:
If we wanted to apply similar logic to all stages, we would do so with a different implementation (likely a preprocessor that could run before task execution).

`selectedStageRoles` may be null, but you're also manipulating the context for every stage with `stageRoles` and `permissions`.

It also introduces another remote call in the flow to run a pipeline (to `front50`) which is even with this feature as currently implemented would be unnecessary to anyone who isn't using roles in manual judgment stages. As a general rule, we try to remove unnecessary rpc calls on critical paths.

I'd like this changed, and really this bit removed.

Reply:
removed this piece of code completely now.